### ISSUE TITLE
Preprocessors

### ISSUE THREAD

**Issue by GilChrist19** [gregbellan/Stabl#23]
Digging through the code, there are 3 preprocessors:

1. [General Preprocessor](https://github.com/gregbellan/Stabl/blob/1d07f85a13cfbecb4f08ce21075bf4fbb8e34678/stabl/multi_omic_pipelines.py#L39)
2. [Testing Preprocessor](https://github.com/gregbellan/Stabl/blob/1d07f85a13cfbecb4f08ce21075bf4fbb8e34678/stabl/multi_omic_pipelines.py#L328)
3. [Validation Preprocessor](https://github.com/gregbellan/Stabl/blob/1d07f85a13cfbecb4f08ce21075bf4fbb8e34678/stabl/multi_omic_pipelines.py#L840C5-L840C17)

I don't quite understand why 2 is different from 1, and why 3 isn't used at all. 

Thank you!

**Comment by xavdurand** [comment]
Hello @GilChrist19 ,

Thank you for your comment.
Firstly, the general preprocessor can also be considered as the training preprocessor. We made it a global variable just for convenience.
Then, the testing and validation preprocessor are the same. One is defined in the CV function, and the other in validation function. It could have been defined as a global variable as well.

Knowing all of that, just to fix a name, let's call the preprocessor 1 `training preprocessor` and the preprocessor 2/3 `validation preprocessor`.

In our current process, for a model like Lasso, the training preprocessor is fitted on the training set and then transform the training and the test set.
In the case of Stabl, the model performs a feature selection. So, the preprocessing is as follow:
1. fit the training preprocessor on the data, then transform the training set
2. fit the Stabl model on the transformed training set to select features
3. reduced the training set to the set of selected features
4. fit and transform the validation preprocessor on the reduced training set
5. transform the test set to reduce to the set of selected features and transform the dataset with the validation preprocessor
6. training of the final model on training set and predict on test set

To put it in a nutshell, the different between the training preprocessor and the validation preprocessor is that we removed the Variance Filter and the Low Info Filter because the variable selection process is already done.

I hope it will answer your question.

Xavier Durand

PS: If it answers your question, I let you close the comment.

**Comment by GilChrist19** [comment]
Morning!

This does help, thank you!
I understand the difference between the `training preprocessor` and the `validation preprocessor`, at least in the context of the CV function.

I still don't follow why the `validation preprocessor` isn't used in `multi_omic_stabl()`, such as [here](https://github.com/gregbellan/Stabl/blob/1d07f85a13cfbecb4f08ce21075bf4fbb8e34678/stabl/multi_omic_pipelines.py#L848) or [here](https://github.com/gregbellan/Stabl/blob/1d07f85a13cfbecb4f08ce21075bf4fbb8e34678/stabl/multi_omic_pipelines.py#L869).

Maybe this relates to another confusion I had (LMK if you'd rather I put this on a new issue) - what's the difference between the `multi_omic_stabl_cv()` and `multi_omic_stabl()` functions? 
It doesn't seem like `multi_omic_stabl()` grabs the trained models output by `multi_omic_stabl_cv()`. 
I know the CV notebooks use the CV function, and the val notebooks use the validation function, but the validation function doesn't seem to do crossvalidation, it just tests on a new dataset. 

Thank you Xavier

**Comment by xavdurand** [comment]
Hi,

Hum, you are totally right. The validation preprocessor is in fact not really used. It is an error, probably the curse of copy-paste haha.
In fact, it should not be a problem as the selected features already passed the preprocessing part.

For the second topic, we decided to split it in two functions, as we would still retrain the model in the validation part.
The only difference between both is that CV is for cross-validation (only on the trainset) and the other one is for validation only (training on trainset and prediction on validation).

Xavier

**Comment by GilChrist19** [comment]
Agreed, need to retrain on all of the data. 

I think my confusion is b/c I can't see where the parameters from my CV are taken into the validation. 
The CV uses a nested-CV (super cool, super impressed here!!), but then none of the examples pass that to the validation. 

I'm guessing I can use the `stabl_params` to transfer the best parameters from the CV function to the validation?

**Comment by xavdurand** [comment]
There is not really a better parameter in Stabl. We are using all the parameter defined in `lambda_grid` or in `self.fitted_lambda_grid_` in the case of auto mode.
This is the power of Stabl, it used all the defined penalties.

Maybe to answer a future question, during a classic process of using Stabl, we determine what is a good range of `lambda_grid` to have a nice looking stability path (output of `plot_stabl_path` [function](https://github.com/gregbellan/Stabl/blob/1d07f85a13cfbecb4f08ce21075bf4fbb8e34678/stabl/stabl.py#L425)). You can also use the `auto` mode as a value of `lambda_grid`. It will determine automatically what is a good range of lambda based on your data using the function [`_get_optimized_lambda_grid`](https://github.com/gregbellan/Stabl/blob/1d07f85a13cfbecb4f08ce21075bf4fbb8e34678/stabl/stabl.py#L1007).

I would say that the CV function is used to have a performance on the dataset so you can change the parameter. After that, you have your final performance on the validation.
